### PR TITLE
Fix transaction blocker

### PR DESCRIPTION
**Pull request by Rados13**

### CODE REVIEW

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/domain/Ledger.kt [hunk 1]
```diff
@@ -59,16 +60,17 @@ data class Ledger(
             lastApplied = it.entry.getId()
         }
 
+//        assert(lastApplied == commitIndex)
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/domain/Ledger.kt:63]: is it still neccessary?

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/domain/Ledger.kt [hunk 2]
```diff
@@ -132,6 +134,7 @@ data class Ledger(
         }
 
     fun isNotApplied(entryId: String): Boolean = !history.containsEntry(entryId)
+    fun isNotAppliedOrProposed(entryId: String): Boolean = !history.containsEntry(entryId) && !proposedEntries.any{it.entry.getId()==entryId}
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/domain/Ledger.kt:137]: Imo `isNotAppliedNorProposed` would be better name, but you don't have to change it

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt [hunk 3]
```diff
@@ -73,7 +74,9 @@ class RaftConsensusProtocolImpl(
 
     private suspend fun sendLeaderRequest() {
         if (executorService != null) {
-            restartTimer(RaftRole.Candidate)
+            mutex.withLock {
```
Comment (Davenury) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt:77]: where did that mutex came from?

**Review by Davenury — APPROVED** [review]

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt (hunk 3 above)
Comment (Rados13) [reply src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt:77]: Here it is initialized
https://github.com/Davenury/Unifying-Consensus-And-Atomic-Commitment-PoC/blob/0f91ed80ee5de1561c76aed8c2617fbf50b6f7ab/src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt#L55

**Review by kjarosh — APPROVED** [review]

**Review by Davenury — APPROVED** [review]

**Pull request merged**